File: app/domain/interfaces/face_detector.py
"""Face detector interface following Interface Segregation Principle."""

from typing import Protocol

import numpy as np

from app.domain.entities.face_detection import FaceDetectionResult


class IFaceDetector(Protocol):
    """Protocol for face detection implementations.

    Implementations can use different algorithms (MTCNN, MediaPipe, RetinaFace, etc.)
    without changing client code (Open/Closed Principle).
    """

    async def detect(self, image: np.ndarray) -> FaceDetectionResult:
        """Detect faces in an image.

        Args:
            image: Input image as numpy array (H, W, C) in BGR format

        Returns:
            FaceDetectionResult containing detection information

        Raises:
            FaceNotDetectedError: When no face is found in the image
            MultipleFacesError: When multiple faces are detected

        Note:
            This method enforces single-face detection. For multi-face detection,
            use a different interface or extend this protocol.
        """
        ...

File: demo-ui/src/lib/biometric/domain/interfaces/face-detector.ts
/**
 * Port for face detection implementations.
 *
 * Mirrors: app/domain/interfaces/face_detector.py → IFaceDetector
 *
 * Implementations can use different algorithms (MediaPipe, TFLite, Server API)
 * without changing client code (Open/Closed Principle).
 */

import type { FaceDetectionResult } from '../entities/face-detection-result';
import type { DetectorInput } from '../types';

export interface IFaceDetector {
  /**
   * Detect a face in an image.
   *
   * @param image - Image data source
   * @returns Face detection result with bounding box and confidence
   * @throws {FaceNotDetectedError} when no face is found
   * @throws {MultipleFacesError} when multiple faces are detected
   */
  detect(image: DetectorInput): Promise<FaceDetectionResult>;

  /** Release resources (WASM memory, GPU textures). */
  dispose(): Promise<void>;
}
